### ISSUE TITLE
[selectors] The forgiving nature of :has breaks jQuery when used with a complex :has selector

### ISSUE THREAD

**Issue by mgol** [w3c/csswg-drafts#7676]
The fact that the native `:has` pseudo-class:
https://drafts.csswg.org/selectors/#relational
takes `forgiving-relative-selector-list` as an argument:
https://drafts.csswg.org/selectors/#forgiving-selector
means the contents are not validated.

jQuery has supported the `:has` pseudo-class for ages. However, its support is more powerful; in particular, it allows for jQuery extensions like `:contains` to appear within `:has`.

The way the jQuery selector engine works (and have worked for a long time), the selector is tried against `querySelectorAll` with minor modifications and if it fails, it goes through the internal selector engine. It's either-or.

Since selectors like `ul:has(li:contains('Item'))` no longer fail in Chrome, jQuery defers to `querySelectorAll` and such a selector returns 0 results even if the jQuery path should match something.

We can of course try to patch it in jQuery, perhaps defaulting selectors containing `:has` to use the jQuery selector engine. But it won't help countless apps using older jQuery versions. So I wanted to bring it for consideration for the standards committee. I understand if the decision is going to be "no changes in the spec are planned" but it'd be good to have this discussion and the decision made knowing the consequences.

As you can see, it didn't take long after the Chrome 105 release for us to get a bug report about this breakage:
https://github.com/jquery/jquery/issues/5098
Implementing `:has` according to the spec makes the browser break the jQuery test suite right now.

**Comment by mgol** [comment]
It's worth noting this affects all jQuery versions from [1.3 released in January, 2009](https://blog.jquery.com/2009/01/14/jquery-1-3-released/) to the current [3.6.1 released a few days ago](https://blog.jquery.com/2022/08/26/jquery-3-6-1-maintenance-release/).

**Comment by lilles** [comment]
This did not break with Safari shipping, because WebKit did not implement using forgiving selector list per spec: https://bugs.webkit.org/show_bug.cgi?id=244708

**Comment by byung-woo** [comment]
Even if we make the `:has()` unforgiving, it seems that there can be still same issues in `:has(:is(.a:has(.b), .c))` since `:is()` is unforgiving. (It seems that @lilles already asked about this here : https://github.com/jquery/jquery/issues/5098#issuecomment-1235404304)

**Comment by anttijk** [comment]
The current WebKit behavior (not by design) is to use forgiving parsing but fail the selector if the selectors lists ends up empty. That is

`:has(:foo, bar) { }`

is valid and parses to

`:has(bar) { }`

but 

`:has(:foo)` is invalid. 

Perhaps this is actually the best option considering the jQuery issue.

**Comment by emilio** [comment]
> The way the jQuery selector engine works (and have worked for a long time), the selector is tried against querySelectorAll with minor modifications and if it fails, it goes through the internal selector engine. It's either-or.

Shouldn't jQuery check `CSS.supports("selector(:has(....))")`, which should be unforgiving per spec? See https://github.com/w3c/csswg-drafts/issues/7280

**Comment by byung-woo** [comment]
> Shouldn't jQuery check `CSS.supports("selector(:has(....))")`, which should be unforgiving per spec? See #7280

I agree that this is the correct way to validate `:has()` usage, but it's sad that even if JQuery applies this to the new version, there may be unexpected regressions when someone uses an older JQuery version. :(

Now I'm trying to merge a fix to Chrome to avoid existing JQuery conflict by following the WebKit's behavior. But I think that we need a way to handle this situation properly.

**Comment by emilio** [comment]
Is there any real-world site broken by the change? How frequent do we expect it to be to use jQuery-specific selectors inside `:has()` to begin with?

**Comment by emilio** [comment]
Also, is this a "recent version of jQuery" kinda thing? Or would this change cause issues on older jQuery versions?

**Comment by emilio** [comment]
If there's no real-world breakage, it might be worth just getting jQuery fixed by using `CSS.supports()`, rather than introduce a weird quirk in how `:has()` works vs. other modern selectors...

**Comment by mgol** [comment]
@emilio 
> Shouldn't jQuery check `CSS.supports("selector(:has(....))")`, which should be unforgiving per spec? See #7280

In Chrome 105, `CSS.supports('selector(:has(div:contains(div)))')` returns `true`. I'd gladly use this in jQuery if this was working as advertised.

> Also, is this a "recent version of jQuery" kinda thing? Or would this change cause issues on older jQuery versions?

As I wrote in https://github.com/w3c/csswg-drafts/issues/7676#issuecomment-1235358708, this affects virtually all jQuery versions used in the wild.

> Is there any real-world site broken by the change? How frequent do we expect it to be to use jQuery-specific selectors inside `:has()` to begin with?

I am unable to answer this question. But it did both break our test suite and get us a report from a user at https://github.com/jquery/jquery/issues/5098. @Rinzwind did you encounter this issue just in manual testing or did it actually break anything for you on a live site?

**Comment by mgol** [comment]
> Now I'm trying to merge a fix to Chrome to avoid existing JQuery conflict by following the WebKit's behavior. But I think that we need a way to handle this situation properly.

Note that the WebKit behavior still breaks some jQuery selectors. True, it doesn't break:
```
div:has(span:contains('Item'))
```
but it still breaks:
```
div:has(div, span:contains('Item'))
```
Maybe that latter kind of selectors is not used as often in the wild.

**Comment by mgol** [comment]
@byung-woo 
> Now I'm trying to merge a fix to Chrome to avoid existing JQuery conflict by following the WebKit's behavior. But I think that we need a way to handle this situation properly.

Note that the WebKit behavior still breaks some jQuery selectors. True, it doesn't break:
```
div:has(span:contains('Item'))
```
but it still breaks:
```
div:has(div, span:contains('Item'))
```
Maybe that latter kind of selectors is not used as often in the wild.

**Comment by lilles** [comment]
Opened issue for CSS.supports()/@supports: https://bugs.chromium.org/p/chromium/issues/detail?id=1359396

**Comment by lilles** [comment]
There is a real-world breakage reported here: https://crbug.com/1359246

**Comment by Rinzwind** [comment]
>  @Rinzwind did you encounter this issue just in manual testing or did it actually break anything for you on a live site?

In our case, none of the selectors used in the application itself seem to be affected by this problem, only the ones in our automated tests (see: https://github.com/jquery/jquery/issues/5098#issuecomment-1233876338).

**Comment by tabatkins** [comment]
Importantly, this is *not* specific to `:has()`, and so any change to `:has()` won't stop the breakage. This breakage is due to the nature of forgiving-selector-list in general, so any place that uses it (like `:is()`, as Byungwoo notes), will trigger the exact same problem - jQuery users putting jQuery-specific selectors into an `:is()` will similarly bypass the current check and stick with the browser's selector behavior. The `:has()`-specific part is just that `:has()` already exists as a jQuery-custom selector, while `:is()` doesn't, so older content using `:has()` will break. That said, new content using `:is()` with older jQuery versions will also continue to get the bad behavior.

So our options are:

1. Rename `:has()` to not overlap with the jQuery name, so older content using older jQuery versions will continue to trigger the custom code path rather than use built-ins.
2. Give up on forgiving-selector-list entirely, switching :is() and :has() to the unforgiving behavior, and never using it again.
3. Accept that older versions of jQuery can cause page breakage, if you use custom jQuery selectors inside of `:has()` (or nest `:has()`, which is allowed in jQuery)

All three options suck! Unless the breakage gets fairly bad, tho, I do lean toward 3. The `:has()` name is *very* nice and already widely socialized, and I'd be extremely loathe to give up on forgiving-selector-list.

**Comment by mgol** [comment]
BTW, re:
> Now I'm trying to merge a fix to Chrome to avoid existing JQuery conflict by following the WebKit's behavior. But I think that we need a way to handle this situation properly.

What use cases exist that make it useful to have `:has` forgiving that **do not** also require not following the WebKit behavior of throwing if none of the `:has` contents use only supported selectors?

**Comment by mgol** [comment]
The fact that in WebKit `qSA('div:has(:foo)')` throws while `qSA('div:has(:foo, span)')` does not is more confusing to me than either not throwing in both cases or throwing in both. It'd be good to learn if there are any arguments for this split to be good for the Web before Chrome jumps in to copy WebKit here and that behavior has to be standardized.

**Comment by mgol** [comment]
@tabatkins 
> jQuery users putting jQuery-specific selectors into an `:is()` will similarly bypass the current check and stick with the browser's selector behavior

Because the jQuery selector engine currently works via a binary switch - first, check qSA; if it throws, use the jQuery custom traversal - and because jQuery never implemented `:is()`, the behavior change is only that before `:is()` got implemented a selector like `:is(div:contains("Foo"))` would throw and once `:is()` got implemented it started returning `0` results. So there's a change but one unlikely to cause a breakage, contrary to the `:has()` situation.

Does Google have any way to collect metrics of how often `:has` is used with jQuery-specific pseudos like `:contains`? That would help a lot.

**Comment by andruud** [comment]
> Does Google have any way to collect metrics of how often :has is used with jQuery-specific pseudos like :contains? That would help a lot.

@mgol We could query HTTP Archive. What are we looking for?

- `:has(.*:contains.*)`
- What else?

**Comment by mgol** [comment]
@andruud 
> We could query HTTP Archive. What are we looking for?
> 
>     * `:has(.*:contains.*)`
> 
>     * What else?

Any jQuery-specific pseudo inside of `:has()`, including:
* positional pseudos: `:first`, `:last`, `:eq`, `:even`, `:odd`, `:lt`, `:gt`, and `:nth`
* other jQuery-specific ones: `:contains`, `:parent`, `:header`, `:input`, `:button`, `:text`, `:radio`, `:checkbox`, `:file`, `:password`, `:image`, `:submit`, `:reset`, `:animated` - I think those are not available natively
* native ones but not available in `:has` natively: `:has` (yes, nested), `:not`

There are other native ones also implemented by jQuery but those are probably fine? `:lang`, `:target`, `:root`, `:focus`, `:enabled`, `:disabled`, `:checked`, `:selected`, `:empty`

**Comment by SelenIT** [comment]
@mgol isn't `:not` available in `:has`? The spec apparently suggests it's allowed (see Example 15). Are there any extra implementation-specific limitations?

**Comment by andruud** [comment]
> 2. Give up on forgiving-selector-list entirely, switching :is() and :has() to the unforgiving behavior, and never using it again.

@tabatkins What about the following?

2b. Don't give up on forgiving-selector-list entirely, but spec the current Safari behavior. I.e. make :has() invalid if _all_ the arguments are invalid.

... provided that it avoids enough problems to be worth it. (As I understand it doesn't fully solve the issue).

**Comment by emilio** [comment]
That seems rather unfortunate fwiw, I'd much rather make them behave consistently regardless of the number of valid arguments...

If we had something like that I think I'd prefer making it unforgiving iff the number of provided arguments is exactly one, rather than making it depend on the number of _valid_ arguments. But still unfortunate IMHO

**Comment by lilles** [comment]
So making:

`div.class:has(:invalid, :invalid) {}`

invalid instead of never matching wouldn't be too bad, but the consequence of dropping other fully valid selectors from a list, which would happen with `main.class` in:

 `main.class, div.class:has(:invalid, :invalid) {}`

is more problematic.

**Comment by mgol** [comment]
@SelenIT 
> @mgol isn't `:not` available in `:has`? The spec apparently suggests it's allowed (see Example 15). Are there any extra implementation-specific limitations?

You are right. What's not supported natively is nested `:has()`, `:not()` in `:has()` seems fine.

I updated the comment.

**Comment by mgol** [comment]
FWIW, I have [a PR to use `CSS.supports("selector(...)")`](https://github.com/jquery/jquery/pull/5107) which I intend to get into `3.6.2`. But, of course, it won't help older versions of jQuery.

**Comment by romainmenke** [comment]
We use `:has(:has(any))` as a feature detection for a polyfill for `querySelectorAll(:has(...))`. This polyfill in turn is used to polyfill `:has` in CSS. (https://github.com/csstools/postcss-plugins)

With Chrome now matching [Safari's (buggy) behaviour](https://github.com/w3c/csswg-drafts/issues/6952#issuecomment-1159279009) we override the native implementation and apply the polyfill.

We obviously don't have the reach of jQuery but it seems important to resolve this asap before too many things depend on the current specification.

------

The current behaviour with a partially forgiving selector is really confusing.
It becomes tricky to know when and how this selector list is valid or invalid :

```css
.anything,
:has(:something-only-supported-from-chrome-200-safari-50-firefox-200) {
  color: green;
}
```

Renaming `:has()` or making it fully unforgiving seems better in the long run while still doing something to prevent breakage.
(old jQuery eventually goes away)

**Comment by mgol** [comment]
@andruud any updates in querying the HTTP Archive?

**Comment by tabatkins** [comment]
Okay, so I gave [three options earlier](https://github.com/w3c/csswg-drafts/issues/7676#issuecomment-1235773031), but in the Sep 16 meeting @fantasai gave a fourth option (maybe an option 2b from my list?) where we keep forgiving-selector-list, but *only* use it in :is() and :where(). Everywhere else we stick to the existing non-forgiving behavior.

* This fixes the question at hand - jQuery will go back to working correctly when an author uses a jquery-specific selector inside of (what they assumes was) jquery-specific :has()
* This makes :has() (and all similar selector arguments, like `:nth-child(of)`) consistent with :not(), which is non-forgiving.
* Forgivingness becomes a special property of :is()/:where(), which seems more learnable than a more arbitrary set of selectors influenced by history.

So I propose we take on fantasai's suggestion.

**Comment by fantasai** [comment]
For the record, I believe this was originally suggested by @dbaron. :)

**Comment by andruud** [comment]
> @andruud any updates in querying the HTTP Archive?

@mgol Unfortunately the "best" upper bound I was able to produce from that was way too high to be actionable. It was also not useful in proving that it definitely _is_ a huge problem, since I wasn't able to even guesstimate how many false positives there were.

But looks like we have an acceptable path forward which doesn't depend on that anyway. (I.e. making everything except :is/:where unforgiving).

**Comment by LeaVerou** [comment]
Seeing this late, but what about slightly changing how `<forgiving-selector-list>` / `<forgiving-relative-selector-list>` so that the whole list if is still invalid if ALL of its selectors are invalid, i.e. not allowing the list to be empty after removing invalid selectors? 

That still allows it to cater to the use cases it was created for, where some selectors in the list are valid and some are not, but would mitigate most (not all) cases of jQuery breakage, and I'd argue it's better DX, as it's more consistent with expectation (I was surprised that a single argument that is invalid did not cause the selector to fail)

**Comment by anttijk** [comment]
I sort of agree with Lea that the current WebKit behavior here is the most sensible one. An empty selector list as a valid case is weird. Note however that only `:has` currently behaves like this, `:is`/`:where` has the spec behavior.

**Comment by LeaVerou** [comment]
> I sort of agree with Lea that the current WebKit behavior here is the most sensible one. An empty selector list as a valid case is weird. Note however that only `:has` currently behaves like this, `:is`/`:where` has the spec behavior.

Good point. I was suggesting we do this for all `<forgiving-selector-list>` / `<forgiving-relative-selector-list>` instances, including in `:is()` and `:where()`.

**Comment by romainmenke** [comment]
> Good point. I was suggesting we do this for all <forgiving-selector-list> / <forgiving-relative-selector-list> instances, including in :is() and :where().

This would be breaking for us.

```css
.foo,
:is(:invalid-in-older-browser, :invalid-in-other-browser, ...) {
  /* styles */
}
```

I mentioned this before : https://github.com/w3c/csswg-drafts/issues/7676#issuecomment-1250070779
The validity of the items in the list is not a constant.
It depends on the browser version.

**Comment by tabatkins** [comment]
@anttijk The problem isn't someone explicitly writing an empty :is(), it's someone writing a normal selector list into the pseudo-class and all of them turning out invalid. If *that* makes *the entire selector* invalid, but having *one* recognized selector keeps it valid, it's a weird effect, as @romainmenke just pointed out.

(Whether an explicitly empty :is() is valid or not is a separate question that I don't particularly care about. We can write the grammar either to allow it or forbid it, but serialization will be trickier if we forbid it, since we drop the invalid selectors from the argument.)

**Comment by anttijk** [comment]
> @anttijk The problem isn't someone explicitly writing an empty :is(), it's someone writing a normal selector list into the pseudo-class and all of them turning out invalid. If _that_ makes _the entire selector_ invalid, but having _one_ recognized selector keeps it valid, it's a weird effect, as @romainmenke just pointed out.

In that example you could include all the selectors inside `:is` list which is probably more natural anyway. But there is not much point talking about it since the ship already sailed. (I don't care about the actually-empty case either).

**Comment by tabatkins** [comment]
Edits checked in, just need to make sure WPT reflects this now.

**Comment by tabatkins** [comment]
And done!

**Comment by AleksandrHovhannisyan** [comment]
Just thinking out loud beyond this issue: Maybe there needs to be more granular versioning for CSS (and HTML) in the future so that authors can opt into new behaviors without user browser versions introducing breaking changes. Imagine if authors could set a meta tag in their document opting into `css3.x.y` and incrementally adopting new features like they would with an open-source library, so long as the user's browser version supports that new feature. This way, an old project running jQuery wouldn't have to worry about naming clashes—it could just lock the CSS version to the one before `:has` was introduced.

(I understand that this is likely a lot harder to achieve in practice than it sounds. I also understand that this clashes with the cross-browser developer experience that we're all familiar with, where features are tied to browser versioning and not author-specified versions.)

The change here was harmless. I'm just not sure how backwards compatible the CSS spec will need to be moving forward so that it doesn't accidentally introduce breaking changes or conflicting behaviors for older, unrelated technologies that were ahead of their time and supported syntaxes that CSS itself did not yet have.
